### ISSUE TITLE
[Feature] [Ascend] Enable mixed batching on the Ascend backend

### ISSUE THREAD

**Issue by Comtive-Wdson** [sgl-project/sglang#12400]
### Checklist

- [x] 1. If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 2. Please use English, otherwise it will be closed.

### Background
Mixed batching is a common and useful technique to improve throughput by batching chunked prefill requests and decode requests together. 

It has not been supported in Ascend backend. The current Ascend backend rejects mixed batches because `torch_npu._npu_flash_attention_qlens` requires sequence lengths to be multiples of 128; decode requests (q_len = 1) violate this alignment and trigger a kernel error.

### My attempt
I have tried to support it by replacing the fixed-length kernel with a new variable-length flash-attention operator `torch_npu.npu_fusion_attention` that accepts arbitrary sequence sizes by setting shape as 'TND' format(https://www.hiascend.com/document/detail/zh/Pytorch/710/apiref/torchnpuCustomsapi/context/torch_npu-npu_fusion_attention.md)

The core code is shown as below (Core snippet only; the rest is the same as before, omitted for brevity.), replacing `torch_npu._npu_flash_attention_qlens` with `torch_npu.npu_fusion_attention` 
```python
#ascend_backend.py
class AscendAttnBackend(AttentionBackend):
	def build_mixed_batch_causal_mask(self, actual_seq_len_q, actual_seq_len_kv):
        max_seq_q = max(actual_seq_len_q)
        max_seq_kv = max(actual_seq_len_kv)
        mask = torch.triu(torch.ones((max_seq_q, max_seq_kv), dtype=torch.bool), diagonal=1)
        return mask.to(self.device)
    
    def forward_extend(
        self,
        q,
        k,
        v,
        layer: RadixAttention,
        forward_batch: ForwardBatch,
        save_kv_cache: bool = True,
    ):
        #...
                # if layer.qk_head_dim <= 128:
                #     query = q.reshape(-1, layer.tp_q_head_num * layer.qk_head_dim)
                #     attn_output = torch.empty(
                #         (query.shape[0], layer.tp_q_head_num * layer.v_head_dim),
                #         dtype=query.dtype,
                #         device=query.device,
                #     )

                #     torch_npu._npu_flash_attention_qlens(
                #         query=query,
                #         key_cache=k_cache,
                #         value_cache=v_cache,
                #         mask=self.mask,
                #         block_table=self.forward_metadata.block_tables,
                #         seq_len=self.forward_metadata.extend_seq_lens_cpu_int,
                #         context_lens=self.forward_metadata.seq_lens_cpu_int,
                #         scale_value=layer.scaling,
                #         num_heads=layer.tp_q_head_num,
                #         num_kv_heads=layer.tp_k_head_num,
                #         out=attn_output,
                #     )
                
                if layer.qk_head_dim <= 128:
                    query = q.reshape(-1, layer.tp_q_head_num, layer.qk_head_dim)
                    key = k_cache.reshape(-1, layer.tp_k_head_num, layer.qk_head_dim)
                    value = v_cache.reshape(-1, layer.tp_v_head_num, layer.v_head_dim)
                    head_num = query.shape[1]
                    if self.forward_metadata.extend_seq_lens_cpu_int is not None:
                        actual_seq_len_q = self.forward_metadata.extend_seq_lens_cpu_int.cpu().int().tolist()
                    else:
                        actual_seq_len_q = None
                    if self.forward_metadata.seq_lens_cpu_int is None:
                        actual_seq_len_kv = self.forward_metadata.seq_lens_cpu_list
                    else:
                        actual_seq_len_kv = (
                            self.forward_metadata.seq_lens_cpu_int.cpu().int().tolist()
                        )
                    attn_mask = self.build_mixed_batch_causal_mask(actual_seq_len_q, actual_seq_len_kv)
                    
                    output = torch_npu.npu_fusion_attention(
                            query,
                            key,
                            value,
                            head_num,
                            pse=None,
                            padding_mask=None,
                            atten_mask=attn_mask,
                            scale=1.0 / math.sqrt(q.shape[-1]),
                            input_layout="TND",
                            actual_seq_qlen=actual_seq_len_q,
                            actual_seq_kvlen=actual_seq_len_kv,
                        )[0]
                    attn_output = output.view(q.shape[0], layer.tp_q_head_num * layer.v_head_dim)
```
### Experiment results
Initial experiments show the operator compiles, the system can serve requests most of time. But sometimes the response is also not the correct answer for the request, and the computed `attn_mask` shape occasionally mismatches the expected [MaxSq, MaxSkv], producing the runtime error when serving some requests:
```shell
get unsupported atten_mask shape, the shape is [565, 14773]. MaxSq=[560], MaxSkv=[14769], when input_layout is TND, the supported atten_mask shape is [MaxSq, MaxSkv].[FUNC:AnalyzeOptionalInput][FILE:flash_attention_score_tiling_general.cpp][LINE:1514] TraceBack (most recent call last):
```

I believe mixed batching is a critical feature; the sketch above outlines my personal attempt and I’d appreciate any corrections, discussions to support the feature.

**Comment by Makcum888e** [comment]
Hi, I also tried using this operator but when using it I faced a drop in accuracy on 20% on the GSM8k dataset https://github.com/sgl-project/sglang/tree/main/benchmark/gsm8k
About usage of this op: 
this op requires key/value in "standart" format ( keys for first question in batch, for second and so on), so it needed to do extracting from cache buffer real values, something like this 
```
per_reqs_tokens = torch.empty(seq_lens_sum,device,dtype)
start = 0
for i in range(forward_batch.req_pool_indices.shape[0]):
    end = start + seq_lens[i]
    per_reqs_tokens[start:end] = forward_batch.req_to_token_pool.req_to_token[req_pool_indices[i],:seq_lens[i])
    start = end
key = k_cache[per_reqs_tokens].view(-1, kv_num_heads, kv_head_size)
```
but while preparation of `per_reqs_tokens` can be done once before run inference in `init_forward_metadata`, this row `key = k_cache[per_reqs_tokens].view(-1, kv_num_heads, kv_head_size)` will have to be executed on each layer and this will lead to a significant drop in performance

And this op also required `actual_seq_qlen` and `actual_seq_kvlen` as a cumsum of real `extend_seq_lens` and `seq_lens`
<img width="1097" height="95" alt="Image" src="https://github.com/user-attachments/assets/ad43377c-7eed-4017-910c-111bc5cb6c31" />

They are also trying to support this feature, but it looks like they also have some problems with accuracy https://github.com/sgl-project/sglang/pull/12491

**Comment by Comtive-Wdson** [comment]
Thank you, I will keep focusing and trying. It is confusing with the accuracy issue you mention.
